internal/raft/node.go

```diff
@@ -438,7 +444,12 @@ func (rc *raftNode) serveChannels() {
 	}
 	rc.confState = snap.Metadata.ConfState
 	rc.snapshotIndex = snap.Metadata.Index
-	rc.appliedIndex = snap.Metadata.Index
+	// Set appliedIndex only if its not already initialised
+	// Note that we also set appliedIndex during init from
+	// storage supplied value.
+	if rc.appliedIndex == 0 {
+		rc.appliedIndex = snap.Metadata.Index
+	}
```

Comment:
Wouldn't this lead to incorrect `appliedIndex` ? Using `snap.Metadata.Index` allows to use the last snapshot index and allows the node to apply the delta of `snap.Metadata.Index - rc.appliedIndex` in the restore stage before starting up the raft engine.

Reply:
This change is only preventing an overwrite of the `appliedIndex` if its set already (see line 126). In other words, if the state machine already has the index for the last applied entry, we use that for `appliedIndex`, else we fallback to using `snap.Metadata.Index`.